drake/systems/plants/@RigidBodyManipulator/doKinematics.m

```diff
@@ -83,7 +84,17 @@
 kinsol.v = v;
 
 if (options.use_mex && model.mex_model_ptr~=0 && isnumeric(q))
-  doKinematicsmex(model.mex_model_ptr,q,options.compute_gradients,v,options.compute_JdotV);
+  if isempty(options.kinematics_cache_ptr_to_use)
+    if options.compute_gradients
+      kinsol.mex_ptr = model.default_kinematics_cache_ptr_with_gradients;
```

Comment:
Fixing things on the Matlab side is not the main objective of this pull request. In fact, I actively tried to avoid making major changes in Matlab, I just want to keep it working as is. Requiring that the caller passes in an existing one means I have to change all usages of doKinematics in Matlab and find a place to store a kinematics cache to reuse. Creating a new DrakeMexPointer to KinematicsCache makes everything much slower.